common/src/currency.rs

```diff
@@ -1,48 +1,131 @@
-use sp_runtime::RuntimeDebug;
+use sp_std::convert::From;
+use sp_std::convert::Into;
+use sp_std::convert::TryFrom;
+use sp_std::convert::TryInto;
+use sp_std::fmt;
+use sp_std::str;
+
+use stellar::types::AssetAlphaNum12;
+use stellar::types::AssetAlphaNum4;
+use stellar::PublicKey;
+use substrate_stellar_sdk as stellar;
 
 use codec::{Decode, Encode};
 
+pub type Bytes4 = [u8; 4];
+pub type Bytes12 = [u8; 12];
+pub type AssetIssuer = [u8; 32];
+
 #[cfg(feature = "std")]
 use serde::{Deserialize, Serialize};
 
-#[derive(Encode, Decode, Eq, PartialEq, Copy, Clone, RuntimeDebug, PartialOrd, Ord)]
+#[derive(Encode, Decode, Eq, PartialEq, Copy, Clone, PartialOrd, Ord)]
 #[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
 pub enum CurrencyId {
     Native,
-    TokenSymbol { code: [u8; 4] },
-}
-
-pub trait StellarAsset {
-    fn asset_code(&self) -> Option<[u8; 4]>;
+    StellarNative,
+    AlphaNum4 { code: Bytes4, issuer: AssetIssuer },
+    AlphaNum12 { code: Bytes12, issuer: AssetIssuer },
 }
 
-impl StellarAsset for CurrencyId {
-    fn asset_code(&self) -> Option<[u8; 4]> {
-        match self {
-            CurrencyId::TokenSymbol { code } => Some(*code),
-            _ => None,
-        }
+impl Default for CurrencyId {
+    fn default() -> Self {
+        CurrencyId::Native
     }
 }
 
-impl CurrencyId {
-    pub fn create_from_slice(slice: &str) -> Self {
+impl TryFrom<(&str, AssetIssuer)> for CurrencyId {
+    type Error = &'static str;
+
+    fn try_from(value: (&str, AssetIssuer)) -> Result<Self, Self::Error> {
+        let slice = value.0;
+        let issuer = value.1;
         if slice.len() <= 4 {
-            let mut code: [u8; 4] = [0; 4];
+            let mut code: Bytes4 = [0; 4];
+            code[..slice.len()].copy_from_slice(slice.as_bytes());
+            Ok(CurrencyId::AlphaNum4 { code, issuer })
+        } else if slice.len() > 4 && slice.len() <= 12 {
+            let mut code: Bytes12 = [0; 12];
             code[..slice.len()].copy_from_slice(slice.as_bytes());
-            CurrencyId::TokenSymbol { code }
+            Ok(CurrencyId::AlphaNum12 { code, issuer })
         } else {
-            panic!("More than 4 bytes not supported")
+            Err("More than 12 bytes not supported")
+        }
+    }
+}
+
+impl From<stellar::Asset> for CurrencyId {
+    fn from(asset: stellar::Asset) -> Self {
+        match asset {
+            stellar::Asset::AssetTypeNative => CurrencyId::StellarNative,
+            stellar::Asset::AssetTypeCreditAlphanum4(asset_alpha_num4) => CurrencyId::AlphaNum4 {
+                code: asset_alpha_num4.asset_code,
+                issuer: asset_alpha_num4.issuer.as_binary().clone(),
```

Comment:
here you can just use `into_binary()` instead of `as_binary().clone()` (same a few lines below)

Reply:
Done